homeassistant/components/picnic/coordinator.py

```diff
@@ -0,0 +1,149 @@
+        cart = self.picnic_api_client.get_cart()
+        last_order = self._get_last_order()
+
+        if not cart or not last_order:
+            raise UpdateFailed("API response doesn't contain expected data.")
+
+        slot_data = self._get_slot_data(cart)
+
+        return {
+            ADDRESS: self._get_address(),
+            CART_DATA: cart,
+            SLOT_DATA: slot_data,
+            LAST_ORDER_DATA: last_order,
+        }
+
+    def _get_address(self):
+        """Get the address that identifies the Picnic service."""
+        if self._user_address is None:
+            address = self.picnic_api_client.get_user()["address"]
+            self._user_address = f'{address["street"]} {address["house_number"]}{address["house_number_ext"]}'
+
+        return self._user_address
+
+    @staticmethod
+    def _get_slot_data(cart: dict) -> dict:
+        """Get the selected slot, if it's explicitly selected."""
+        selected_slot = cart.get("selected_slot", {})
+        available_slots = cart.get("delivery_slots", [])
+
+        if selected_slot.get("state") == "EXPLICIT":
+            slot_data = filter(
+                lambda slot: slot.get("slot_id") == selected_slot.get("slot_id"),
+                available_slots,
+            )
+            if slot_data:
+                return next(slot_data)
+
+        return {}
+
+    def _get_last_order(self) -> dict:
+        """Get data of the last order from the list of deliveries."""
+        # Get the deliveries
+        deliveries = self.picnic_api_client.get_deliveries(summary=True)
+        if not deliveries:
+            return {}
+
+        # Determine the last order
+        last_order = copy.deepcopy(deliveries[0])
+
+        #  Get the position details if the order is not delivered yet
+        delivery_position = {}
+        if not last_order.get("delivery_time"):
+            try:
+                delivery_position = self.picnic_api_client.get_delivery_position(
+                    last_order["delivery_id"]
+                )
+            except ValueError:
+                # No information yet can mean an empty response
+                pass
+
+        # Determine the ETA, if available, the one from the delivery position API is more precise
+        # but it's only available shortly before the actual delivery.
+        last_order["eta"] = delivery_position.get(
+            "eta_window", last_order.get("eta2", {})
+        )
+
+        # Determine the total price by adding up the total price of all sub-orders
+        total_price = 0
+        for order in last_order.get("orders", []):
+            total_price += order.get("total_price", 0)
+
+        # Sanitise the object
+        last_order["total_price"] = total_price
+        last_order.setdefault("delivery_time", {})
+        if "eta2" in last_order:
+            del last_order["eta2"]
+
+        # Make a copy because some references are local
+        return last_order
+
+    async def _update_auth_token(self):
```

Comment:
This method doesn't need to be async.
```suggestion
    @callback
    def _update_auth_token(self):
```